Finish GraphCache implementation for Pub/Sub

Yadunund opened the pull request:
This PR:
- Refactors how we store and update the `GraphCache` within each session. Specifically we move away from a `YAML::Node` to an `unordered_map` (although implementing a conversion from our custom map to `YAML` may be helpful for logging or transmitting the graph as a serialized `YAML` over the network).
- Makes the switch to rely on liveliness tokens to broadcast changes to the ROS graph. The previous implementation of relying on put/del/get convey the changes is retained but commented out as it could be reused to support discovery with `zenoh-pico`.
- Implements some basic topic/node introspection functions in `rmw` (the rest will be done in a separate PR).

> Note: Without the router, nodes will not be able to discover each other since multicast discovery is disabled by default in the node's session config. Instead, nodes will receive discovery information about other peers via the router's gossip function.

Yadunund committed:
Add storage plugin config to router config

Signed-off-by: Yadunund <yadunund@openrobotics.org>

Yadunund committed:
Fix non empty node namespace

Signed-off-by: Yadunund <yadunund@openrobotics.org>

Yadunund committed:
Rely on unordered_map instead of yaml for graph cache

Signed-off-by: Yadunund <yadunund@openrobotics.org>

Yadunund committed:
Update graph cache with publisher data

Signed-off-by: Yadunund <yadunund@openrobotics.org>

Yadunund committed:
Implement basic publisher introspection

Signed-off-by: Yadunund <yadunund@openrobotics.org>

Yadunund committed:
Switch to liveliness tokens (#67)

* Switch to liveliness tokens

Signed-off-by: Yadunund <yadunund@openrobotics.org>

* Use zc APIs instead of macros to resolve liveliness api issues

Signed-off-by: Yadunund <yadunund@openrobotics.org>

---------

Signed-off-by: Yadunund <yadunund@openrobotics.org>

Yadunund committed:
Cleanup yaml-cpp

Signed-off-by: Yadunund <yadunund@openrobotics.org>

Yadunund committed:
Cleanup previous graph cache impl

Signed-off-by: Yadunund <yadunund@openrobotics.org>

Yadunund committed:
Refactor topic cache

Signed-off-by: Yadunund <yadunund@openrobotics.org>

Yadunund committed:
Implement liveliness tokens for subscriptions and update graph

Signed-off-by: Yadunund <yadunund@openrobotics.org>

francocipollone commented on rmw_zenoh_cpp/src/detail/graph_cache.cpp:
```diff
@@ -44,8 +45,17 @@ static std::string generate_base_token(
   std::stringstream token_ss;
-  // TODO(Yadunund): Empty namespace will contain /. Fix non-empty namespace.
-  token_ss << "@ros2_lv/" << domain_id << "/" << entity << namespace_ << name;
+  token_ss << "@ros2_lv/" << domain_id << "/" << entity << namespace_;
```
The use of @ is to indicate that it is for internal use? (I know that we are using the same approach to zenoh-dds plugin (according to slides) however I don't remember what's this for) Adding a comment with that would be nice for future reviews.

francocipollone replied:
As it is being used in more than one method we could create a static const variable with that definition and replace all the ocurrences

Yadunund replied:
Just a custom admin space for liveliness tokens https://zenoh.io/docs/manual/abstractions/#admin-space

francocipollone commented on rmw_zenoh_cpp/src/detail/graph_cache.cpp:
```diff
@@ -44,8 +45,17 @@ static std::string generate_base_token(
-  // TODO(Yadunund): Empty namespace will contain /. Fix non-empty namespace.
-  token_ss << "@ros2_lv/" << domain_id << "/" << entity << namespace_ << name;
+  token_ss << "@ros2_lv/" << domain_id << "/" << entity << namespace_;
+  // An empty namespace from rcl will contain "/"" but zenoh does not allow keys with "//".
```
nit: extra "

```suggestion
  // An empty namespace from rcl will contain "/" but zenoh does not allow keys with "//".
```

francocipollone commented on rmw_zenoh_cpp/src/detail/graph_cache.cpp:
```diff
@@ -44,8 +45,17 @@ static std::string generate_base_token(
+  // An empty namespace from rcl will contain "/"" but zenoh does not allow keys with "//".
+  // Hence we add an "_" to denote an empty namespace such that splitting the key
+  // will always result in 5 parts.
+  if (namespace_ == "/") {
+    token_ss << "_/";
+  } else {
+    token_ss << "/";
+  }
```
nit: it can be repaced by one line:

```suggestion
  token_ss << (namespace_ == "/") ? "_/" : "/";
```

francocipollone replied:
nit: I think that a docstring for the method with the expected token format will be neat

clalancette replied:
While this could be a one-liner, I actually think it is harder to read that way.  So I left it as-is.

francocipollone commented on rmw_zenoh_cpp/src/detail/graph_cache.cpp:
```diff
@@ -69,6 +79,22 @@ std::string GenerateToken::publisher(
 {
   std::string token = generate_base_token("MP", domain_id, node_namespace, node_name);
   token += topic + "/" + type + "/" + qos;
+  printf("GenerateToken::Publisher: %s\n", token.c_str());
```
nit: printf left behind. Perhaps it could be evolved to a debug message or just be removed

francocipollone commented on rmw_zenoh_cpp/src/detail/graph_cache.cpp:
```diff
@@ -69,6 +79,22 @@ std::string GenerateToken::publisher(
+{
+  std::string token = generate_base_token("MS", domain_id, node_namespace, node_name);
+  token += topic + "/" + type + "/" + qos;
+  printf("GenerateToken::Subscription: %s\n", token.c_str());
```
nit: same comment here

francocipollone commented on rmw_zenoh_cpp/src/detail/graph_cache.cpp:
```diff
@@ -270,33 +191,185 @@ static std::vector<std::string> split_keyexpr(const std::string & keyexpr)
+    entity != "MS" &&
+    entity != "SS" &&
+    entity != "SC")
+  {
```
We should document the entities somewhere in the code:
```
NN = node?
MP: = Publisher
MS: = Subscriber
SS: = Service Server
SC: = Service Client
```

clalancette replied:
I documented this in the docblock for `generate_base_token`.

francocipollone commented on rmw_zenoh_cpp/src/detail/graph_cache.cpp:
```diff
@@ -270,33 +191,185 @@ static std::vector<std::string> split_keyexpr(const std::string & keyexpr)
+    } else if (entity == "MS") {
+      node.subs.push_back(std::move(data));
+    } else {
+      // TODO.
```
```suggestion
      RCUTILS_LOG_ERROR_NAMED(
        "rmw_zenoh_cpp",
        "Entity type is not supported yet");
```

francocipollone commented on rmw_zenoh_cpp/src/detail/graph_cache.cpp:
```diff
@@ -270,33 +191,185 @@ static std::vector<std::string> split_keyexpr(const std::string & keyexpr)
+    auto ns_it = graph_.find(node->ns);
+    if (ns_it == graph_.end()) {
+      // New namespace.
+      std::string ns = node->ns;
```
nit: only used below, this extra line can be avoided

francocipollone commented on rmw_zenoh_cpp/src/detail/graph_cache.cpp:
```diff
@@ -270,33 +191,185 @@ static std::vector<std::string> split_keyexpr(const std::string & keyexpr)
+      };
+      graph_.insert(std::make_pair(std::move(ns), std::move(map)));
+    } else {
+      auto insertion = ns_it->second.insert(std::make_pair(node->name, node));
+      if (!insertion.second) {
```
We can rely on `std::find_if` from `<algorithm>` here as we are in cpp 17

Yadunund replied:
`std::find_if()` will have O(n) complexity while the insertion approach should have `O(log n)`. The idea is that insertion will fail if the entry exists and will give us an iterator to the element that already exists which can be updated.

Yadunund replied:
Rust has a nicer way to do this with HashMap's`entry().or_insert()` [pattern](https://stackoverflow.com/questions/30414424/how-can-i-update-a-value-in-a-mutable-hashmap).

francocipollone commented on rmw_zenoh_cpp/src/detail/graph_cache.cpp:
```diff
@@ -312,29 +385,121 @@ void GraphCache::parse_put(const std::string & keyexpr)
+        // Here we iterate throught the list of publishers and remove the one
+        // with matching name, type and qos.
+        // TODO(Yadunund): This can be more optimal than O(n) with some caching.
+        auto erase_it = found_node->pubs.begin();
+        for (; erase_it != found_node->pubs.end(); ++erase_it) {
+          const auto & pub = *erase_it;
+          if (pub.topic == node->pubs.at(0).topic &&
+            pub.type == node->pubs.at(0).type &&
+            pub.qos == node->pubs.at(0).qos)
+          {
+            break;
+          }
+        }
```
nit: use `find_if` here and there 
```suggestion
        auto erase_it = std::find_if(found_node->pubs.begin(), found_node->pubs.end(), [&node](const auto &pub) {
            return pub.topic == node->pubs.at(0).topic &&
                   pub.type == node->pubs.at(0).type &&
                   pub.qos == node->pubs.at(0).qos;
        });
```
Then proceed as you do:

```cpp
        if (erase_it != found_node->pubs.end()) {
            // Found a matching element, erase it
            found_node->pubs.erase(erase_it);
            //...
            //...
            //...
        }
```

francocipollone commented on rmw_zenoh_cpp/src/detail/graph_cache.cpp:
```diff
@@ -370,7 +535,14 @@ rmw_ret_t GraphCache::get_node_names(
+  for (auto it = graph_.begin(); it != graph_.end(); ++it) {
+    nodes_number += it->second.size();
+  }
+  // TODO(Yadunund): Delete.
```
Is this TODO for the logging?

francocipollone commented on rmw_zenoh_cpp/src/rmw_get_topic_names_and_types.cpp:
```diff
@@ -31,63 +33,69 @@ rmw_get_topic_names_and_types(
-        RMW_SAFE_FWRITE_TO_STDERR("failed to cleanup names and types during error handling");
-      }
-    });
+  // rmw_ret_t ret = rmw_names_and_types_init(topic_names_and_types, 1, allocator);
+  // if (ret != RMW_RET_OK) {
```
nit: Missing removing commented code

francocipollone commented on rmw_zenoh_cpp/src/detail/graph_cache.cpp:
```diff
@@ -423,31 +595,30 @@ rmw_ret_t GraphCache::get_node_names(
+          return RMW_RET_BAD_ALLOC;
+        }
+      }
+      ++j;
```
General comment for this method. 
I wasn't able to retrieve the list of nodes via `ros2 node list`. I verified and we de call this method from `rmw_get_node_names`. I couldn't debug yet to see what is going on.

Yadunund replied:
Surprised to hear this. Was the router running? Without the router the node that gets spun via `ros2 node list` will not get gossip information about other nodes.

clalancette replied:
This ended up working for me, as long as the router is running.  @francocipollone give it another shot with the latest, see if you can make it work.

francocipollone replied:
Confirmed --> https://github.com/ros2/rmw_zenoh/pull/66#issuecomment-1817112633

francocipollone reviewed (commented):
Great Job @Yadunund  :rocket: 
I left a bunch of not-game-changing comments, ptal

clalancette committed:
Fix types and type mangling.

When playing around with this I found that the types
weren't being properly filled in, and also weren't being
demangled.  Fix both of those things here, as well
as do a lot of cleanup.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette commented:
I ended up rebasing this on the latest, as well as fixing some bugs that I found.  I also addressed most of @francocipollone 's comments, with the exception of a switch to `find_if`.  I'll take a closer look at that later.

However, with all of that done, `ros2 node list` and `ros2 topic list -t` both work entirely correctly for me.  There is still an open TODO in there where a topic with multiple types on it won't be shown properly.  To fix that, we'll need to revamp the way some of the data is put together.  That's a project for next week.

francocipollone commented:
> I ended up rebasing this on the latest, as well as fixing some bugs that I found. I also addressed most of @francocipollone 's comments, with the exception of a switch to `find_if`. I'll take a closer look at that later.
> 
> However, with all of that done, `ros2 node list` and `ros2 topic list -t` both work entirely correctly for me. There is still an open TODO in there where a topic with multiple types on it won't be shown properly. To fix that, we'll need to revamp the way some of the data is put together. That's a project for next week.

Confirmed, it is working correctly. It seems it was my bad: I had some issues with the `daemon`. Basically, it was already running (I was working on another branch) and then when I checkouted this branch and built I hadn't stopped the daemon.

Yadunund commented:
Thanks for cleaning this up Chris! I'll work on massaging the cache data structures to support multiple types on the same topic in a separate PR to keep the diffs minimal. If both of you are okay with the code here, should we merge this in first?

Yadunund committed:
Update README

Signed-off-by: Yadunund <yadunund@openrobotics.org>
